api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java

```diff
@@ -0,0 +1,167 @@
+import org.pmiops.workbench.compliance.ComplianceService;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.db.dao.UserDao;
+import org.pmiops.workbench.db.dao.UserService;
+import org.pmiops.workbench.db.model.DbUser;
+import org.pmiops.workbench.exceptions.ForbiddenException;
+import org.pmiops.workbench.firecloud.FireCloudService;
+import org.pmiops.workbench.google.DirectoryService;
+import org.pmiops.workbench.test.FakeClock;
+import org.pmiops.workbench.test.FakeLongRandom;
+import org.pmiops.workbench.testconfig.UserServiceTestConfiguration;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.Scope;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+public class RasLinkServiceTest {
+  private static final Timestamp NOW = Timestamp.from(Instant.now());
+  private static final FakeClock CLOCK = new FakeClock(NOW.toInstant(), ZoneId.systemDefault());
+
+  private static final String AUTH_CODE = "code";
+  private static final String REDIRECT_URL = "url";
+  private static final String ACCESS_TOKEN = "access_token_1";
+  private static final String LOGIN_GOV_USERNAME = "foo@gmail.com";
+  private static final String USER_INFO_JSON_LOGIN_GOV =
+      "{\"preferred_username\":\"user1@Login.Gov.com\", \"email\":\"" + LOGIN_GOV_USERNAME + "\"}";
+  private static final String USER_INFO_JSON_ERA =
+      "{\"preferred_username\":\"user2@eraCommons.com\", \"email\":\"" + LOGIN_GOV_USERNAME + "\"}";
+  private static final String ID_TOKEN_JWT_IAL_1 =
+      JWT.create()
+          .withClaim(
+              ACR_CLAIM,
+              "https://stsstg.nih.gov/assurance/ial/1 https://stsstg.nih.gov/assurance/aal/1")
+          .sign(Algorithm.none());
+  private static final String ID_TOKEN_JWT_IAL_2 =
+      JWT.create()
+          .withClaim(
+              ACR_CLAIM,
+              "https://stsstg.nih.gov/assurance/ial/2 https://stsstg.nih.gov/assurance/aal/1")
+          .sign(Algorithm.none());
+  private static final TokenResponse TOKEN_RESPONSE_IAL1 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_1);
+  private static final TokenResponse TOKEN_RESPONSE_IAL2 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_2);
+
+  private long userId;
+  private static DbUser currentUser;
+
+  private RasLinkService rasLinkService;
+  private ObjectMapper objectMapper = new ObjectMapper();
+
+  @Autowired private UserService userService;
```

Comment:
A few things that make this test challenging:

1. UserService is a grab-bag of a bunch of random stuff, so it has a bunch of divergent dependencies.
2. you are instantiating a real implementation of UserService (not a mock). This makes for a better test, but also means you need to provide it with its dependencies. Personally, this also the approach I would probably also take.

I'm not sure any of this gets easier with another version of Spring boot. We could setup a partial config which mocks out all the deps of `UserService`, similar to `UserServiceTestConfiguration`, but I don't think it's a great idea to push mock definitions outside of the test itself. I think in an ideal world, there'd be a "testconfig" which worked without mocks, i.e. using fake in-memory implementations. This requires some work to build and scope out, though.

Reply:
Now I feel I use UserService wrongly. I am looking for a layer between `RasLinkService` and `UserDao`, which handles the extra process when updating/reading from Db. What I really need is `updateUserWithRetries` with the function that also update other tables, also `Agent` stuff.  I can try to let RasService directly use `UserDao` with some extra logic, but I feel that is harder. 
What we can do is 
(1) split UserService into `UserStorageService` and `UserService`. 
(2) Further improve: pull those random stuff out into a small service. for example, like the RAS, put extra dependencies(OidcClietn) in its own service, and only use UserStorageService to do the final pure DB stuff. 

Step 1 might be a good start that we can discuss with team.